design/splitfrr-proposal.md

```diff
@@ -0,0 +1,465 @@
+
+- A user may want to connect to another neighbor
+- A user may want to advertise an extra set of IPs for a given neighbor (i.e. the pod’s CIDR)
+- A user may want to override the configuration put in place by MetalLB to reject the incoming routes
+
+In order to provide an abstraction of the FRR Configuration, we need to represent the following entities:
+
+The router:
+
+- ASN
+- ID
+- VRF
+
+For each router, we must specify a neighbor with all the details we have in a session:
+
+- IP
+- ASN
+- passwords
+- bfd profile (if enabled)
+- the ips we want to advertise
+
+For each neighbor we must specify the route-map entries, and in particular:
+
+- the list of prefixes we allow FRR to advertise
+- the list of prefixes we allow FRR to receive
+- the list of prefixes we would set the community to
+- the list of prefixes we would set a localpref to when announcing
+
+If we split the configuration in multiple sub-entities (router, neighbor for router, allowed ips for any neighbor), the configuration of a single node becomes spread across multiple rows of multiple entities.
+
+Because of this, a single CRD with substructures is going to provide better
+readability and it's going to be easier to reason about.
+
+The spec side of a FRRConfiguration would look like (note that this might be subject to changes while proceeding with the implementation):
+
+```go
+
+type FRRConfigurationSpec struct {
+  BGP          BGPConfig `json:"bgp,omitempty"`
+  NodeSelector map[string]string `json:"nodeselector,omitempty"`
+  RawConfig    string `json:"raw,omitempty"`
+}
+
+type BGPConfig struct {
+  Routers []Router `json:"routers"`
+  BFDProfiles []BFDProfile `json:"bfdProfiles,omitempty"`
+}
+
+// Router represent a neighbor router we want FRR to connect to
+type Router struct {
+  ASN uint32 `json:"asn"`
+  ID string `json:"id,omitempty"`
+  VRF string `json:"vrf,omitempty"`
+  Neighbors []Neighbor `json:"neighbors,omitempty"`
+  PrefixesIPV4 []string `json:"prefixesIpV4,omitempty"`
+  PrefixesIPV6 []string `json:"prefixesIpV6,omitempty"`
+}
+
+type Neighbor struct {
+  Address string `json:"address"`
+  Port uint16 `json:"port,omitempty"`
+  PasswordSecret v1.SecretReference `json:"password,omitempty"`
+  HoldTime metav1.Duration `json:"holdTime,omitempty"`
+  KeepaliveTime metav1.Duration `json:"keepaliveTime,omitempty"`
+  EBGPMultiHop bool `json:"ebgpMultiHop,omitempty"`
+  BFDProfile string `json:"bfdProfile,omitempty"`
+  ToAdvertise Advertise `json:"toAdvertise,omitempty"`
+  ToReceive Receive `json:"toReceive,omitempty"`
+}
+
+type Advertise struct {
+  AllowedPrefixes `json:"allowed,omitempty"`
+  PrefixesWithLocalPref []LocalPrefPrefixes `json:"withLocalPref,omitempty"`
+  PrefixesWithCommunity []CommunityPrefixes `json:"withCommunity,omitempty"`
+}
+
+type Receive struct {
+  AllowedPrefixes `json:"allowed,omitempty"`
+}
+
+type AllowedPrefixes struct {
+  Prefixes []string `json:"prefixes,omitempty"`
+  AllowAll bool     `json:"allowAll,omitempty"`
```

Comment:
I assume it would be detailed later, but what happens when allowAll=true and prefixes are defined (do the prefixes decide?) pre/post merging configurations?

Reply:
The spirit of the API is that configurations can be extended but not removed. In that sense, allowall is broader then a set of prefixes so it will force all the prefixes to be sent / received.
Added a note about it.